### PR TITLE
Add determinant and sign-log-determinant functions to mlx.core.linalg

### PR DESCRIPTION
**Pull request by abhilashreddys**
## Proposed changes: Add `linalg.det` and `linalg.slogdet`

Related: https://github.com/ml-explore/mlx/issues/3335

Add determinant (`det`) and sign-log-determinant (`slogdet`) functions to `mlx.core.linalg`.

- `mx.linalg.det(a)` - returns the determinant of a square matrix
- `mx.linalg.slogdet(a)` - returns `(sign, logabsdet)` for numerical stability

### Composite on `lu_factor()`, no new primitive

Both functions compose on the existing `lu_factor()` primitive rather than introducing a new one. This mirrors how JAX implements `slogdet` and is consistent with how `solve()` already builds on `lu()` in this codebase. 

### `slogdet` as the core, `det` derived from it

`slogdet` is the numerically stable. For the general case (N > 3), `det` simply calls `slogdet` and reconstructs via `sign * exp(logabsdet)`. This avoids duplicating the LU-based logic. An internal `slogdet_impl` function handles the shared logic so the public `det` and `slogdet` entry points each validate and promote types exactly once.

### Small-matrix fast paths (1x1, 2x2, 3x3)

A shared `det_raw_small()` helper computes determinants directly using closed-form formulas (element extraction via slice/squeeze, then the standard cofactor expansion for 3x3). This avoids LU overhead for the small matrices common in batched ML workloads (e.g., 3x3 covariance matrices in normalizing flows).

`det` calls `det_raw_small` directly for N <= 3, skipping the log/exp round-trip entirely. `slogdet` calls it too, then splits the raw result into `sign(raw)` and `log(abs(raw))`.

### Integer auto-promotion

`det`/`slogdet` promote integer inputs to float32 via `at_least_float()`. This matches NumPy's behavior (`np.linalg.det` accepts integer arrays).

### Singular matrix handling

For the LU path, singularity is detected by checking for exact zeros on the U diagonal (same approach as NumPy). Singular matrices return `(sign=0, logabsdet=-inf)` from `slogdet`, and `0.0` from `det`.

Thanks @Lucas-Fernandes-Martins for pointing out the singular matrix [issue](https://github.com/ml-explore/mlx/issues/3335#issuecomment-4255476950) in LU factorization. 

## Testing

- **Correctness**: Compared against `np.linalg.det` / `np.linalg.slogdet` for 1x1 through 5x5, random matrices, identity
- **Batched**: Shapes `(3, 4, 4)` and `(2, 3, 3, 3)` match NumPy
- **Numerical stability**: `0.1 * I_100` (det underflows in float32, slogdet stays finite)
- **Singular matrices**: `sign=0`, `logabsdet=-inf`
- **Consistency**: `sign * exp(logabsdet)` matches `det` for non-singular inputs
- **Dtypes**: float32, float64, integer auto-promotion
- **Error handling**: non-square and 1D inputs raise `ValueError`
- **Full suite**: All 18 existing linalg tests continue to pass (186 subtests)

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/linalg.cpp [hunk 1]
```diff
@@ -705,4 +705,162 @@ array solve_triangular(
   return matmul(a_inv, b, s);
 }
 
-} // namespace mlx::core::linalg
\ No newline at end of file
+void validate_det(
+    const array& a,
+    const StreamOrDevice& stream,
+    const std::string& fname) {
+  check_cpu_stream(stream, fname);
```
Comment (zcbenz) [inline mlx/linalg.cpp:712]: No need to check cpu stream since `lu_factor` already does it, and there are code paths not requiring `lu_factor`.

**Review by zcbenz — APPROVED** [review]
Looks good to me, thanks!

Can you rebase the branch to fix the flaky test?

**Comment by zcbenz** [comment]
Also ping @abeleinin who wrote https://github.com/ml-explore/mlx/pull/1451 for a review if possible.

File: mlx/linalg.cpp (hunk 1 above)
Comment (abhilashreddys) [reply mlx/linalg.cpp:712]: Fixed it and also did the rebase.

Comment (abeleinin) [reply mlx/linalg.cpp:712]: @zcbenz Maybe `validate_det` should have a more explicit throw since it supports GPU for smaller sizes (unlike the other linalg ops):
```
[linalg::det] GPU support is currently only available for matrices up to 3x3.
```
Also, this functionality should be noted in the python doc string.

File: mlx/linalg.cpp [hunk 2]
```diff
@@ -705,4 +705,157 @@ array solve_triangular(
   return matmul(a_inv, b, s);
 }
 
-} // namespace mlx::core::linalg
\ No newline at end of file
+void validate_det(const array& a, const std::string& fname) {
```
Comment (abeleinin) [inline mlx/linalg.cpp]: Add `check_float(...)` inside this validate since `mx.complex` isn't supported in `n > 3` case

File: python/tests/test_linalg.py [hunk 3]
```diff
@@ -616,6 +629,210 @@ def test_solve_triangular(self):
         expected = np.linalg.solve(a, b)
         self.assertTrue(np.allclose(result, expected))
 
+    def test_det(self):
```
Comment (abeleinin) [inline python/tests/test_linalg.py:632]: Add a test case for empty matrix `mx.linalg.det(mx.zeros((0, 0)))`

Numpy returns `np.linalg.det(np.zeros((0, 0))) = 1.0` and `np.linalg.slogdet(np.zeros((0, 0))) = (1.0, 0.0)`, which is just the empty product.

Right now `validate_det` will accept this square matrices, but `det_raw_small` falls into the 3x3 branch for `n == 0`. We probably need an explicit `n == 0` case in `det_raw_small`.

**Review by abeleinin — COMMENTED** [review]
PR looks good! Added some comments about edge cases. @zcbenz Feel free to veto any of my suggestions.

File: mlx/linalg.cpp (hunk 1 above)
Comment (zcbenz) [reply mlx/linalg.cpp:712]: This is a very good point, but I think it would be weird if an op defaults to GPU for some shapes but to CPU otherwise, we should probably just constrain to CPU stream only for simplicity, i.e. bringing the check back. Sorry for bouncing back and forth!

File: mlx/linalg.cpp (hunk 2 above)
Comment (abhilashreddys) [reply mlx/linalg.cpp]: Added, using `issubdtype(..., complexfloating)` so int auto-promotion still works. Added regression tests for both the small-matrix and LU paths.

File: mlx/linalg.cpp (hunk 1 above)
Comment (abhilashreddys) [reply mlx/linalg.cpp:712]: Done, moved check_cpu_stream back into `validate_det`. By going CPU-only across all sizes, the explicit GPU error and the docstring note aren't needed, since the behavior is now uniform with the rest of `mlx::linalg`.

File: python/tests/test_linalg.py (hunk 3 above)
Comment (abhilashreddys) [reply python/tests/test_linalg.py:632]: added n == 0 short-circuit in `det_raw_small` returning 1.0 (empty product) broadcast to the batch shape. This fixes both `det` and `slogdet` since `slogdet_impl` delegates to `det_raw_small` for n <= 3. 

Added tests for scalar and batched empty inputs.
